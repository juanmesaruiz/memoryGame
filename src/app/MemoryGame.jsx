import React, { useEffect, useState } from "react";
import { connect } from "react-redux";

import {
  memoryGameClear,
  memoryGameToggleDemonstration
} from "../actions/memoryGameActions";

import {
  getMemoryGameCpuCardSequence,
  getMemoryGameGameCards,
  getMemoryGameIsGameDemonstration
} from "../reducers";

import MemoryGameActions from "./MemoryGameActions";
import MemoryGameCard from "./MemoryGameCard";
import MemoryGameScore from "./MemoryGameScore";

import { arrayHSCards } from "../config/constants";

import "./memoryGame.scss";

const MemoryGame = props => {
  const [activeCard, setActiveCard] = useState(null);

  const {
    cpuCardSequence,
    gameCards,
    isGameDemonstration,
    memoryGameClear,
    memoryGameToggleDemonstration
  } = props;

  const finnishDemonstration = () => {
    memoryGameToggleDemonstration();
  };

  useEffect(() => {
    if (isGameDemonstration) {
      setTimeout(() => {
        cpuCardSequence.forEach((el, i) => {
          setTimeout(() => {
            setActiveCard(null);
            setActiveCard(el);
            if (i === cpuCardSequence.length - 1) finnishDemonstration();
          }, 1250 * i);
        });
      }, 500);
    }
  }, [isGameDemonstration]);

  useEffect(() => {
    return () => {
      memoryGameClear();
    };
  }, [memoryGameClear]);

  const actualCards = arrayHSCards.slice(0, gameCards);

  return (
    <article className="memoryGame">
      <section className="memoryGame-cards">
        {actualCards.map((element, i) => (
          <MemoryGameCard
            activeCard={activeCard}
            idHsCard={arrayHSCards[i].idHsCard}
            id={i + 1}
            key={arrayHSCards[i].idHsCard}
          />
        ))}
      </section>
      <section className="memoryGame-action">
        <MemoryGameActions />
        <MemoryGameScore />
      </section>
    </article>
  );
};

const mapStateToProps = state => {
  return {
    cpuCardSequence: getMemoryGameCpuCardSequence(state),
    gameCards: getMemoryGameGameCards(state),
    isGameDemonstration: getMemoryGameIsGameDemonstration(state)
  };
};

export default connect(
  mapStateToProps,
  { memoryGameClear, memoryGameToggleDemonstration }
)(MemoryGame);
