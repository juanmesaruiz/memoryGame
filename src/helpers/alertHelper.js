export const alertUserError = bestScore => {
  alert(`Ohh! :( You failed! Try it again and beat your best score! (${bestScore})`)
}
