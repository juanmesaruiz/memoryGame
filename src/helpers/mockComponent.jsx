import React from 'react';

export const mockComponent = (name, props) => {
  const { children, ...restOfProps } = props;
  return (
    <code name={name} props={restOfProps}>
      {children}
    </code>
  );
};
