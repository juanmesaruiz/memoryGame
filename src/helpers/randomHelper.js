import { minNumberRandom } from "../config/constants";

export const getRandom = (max, min = minNumberRandom) => {
  return Math.floor(Math.random() * (max - min)) + min;
};
