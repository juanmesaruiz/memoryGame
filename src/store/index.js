import { applyMiddleware, compose, createStore } from 'redux'
import reducers from '../reducers'
import thunk from 'redux-thunk'

const configureStore = () => {
  const initialState = {}
  const devToolsExtension = window.__REDUX_DEVTOOLS_EXTENSION__
  const enhancers = [devToolsExtension()]

  const composedEnhancers = compose(
    applyMiddleware(thunk),
    ...enhancers
  )

  const store = createStore(reducers, initialState, composedEnhancers)

  return { store }
}

export default configureStore
