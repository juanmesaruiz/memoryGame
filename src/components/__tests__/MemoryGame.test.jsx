import React from 'react'
import ReactDOM from 'react-dom'
import { cleanup, render } from '@testing-library/react'
import { Provider } from 'react-redux'

import {
  getMemoryGameCpuCardSequence,
  getMemoryGameGameCards,
  getMemoryGameIsGameDemonstration
} from '../../reducers'

import store from '../../store/__mocks__/mockStore'
import { mockComponent } from '../../helpers/mockComponent'

import MemoryGame from '../MemoryGame'

jest.mock('../../reducers')
jest.mock('../MemoryGameCard', () => {
  return props => mockComponent('MemoryGameCard', props)
})

const component = (
  <Provider store={store}>
    <MemoryGame />
  </Provider>
)

describe('MemoryGame test suite', () => {
  beforeEach(() => {
    getMemoryGameCpuCardSequence.mockReturnValue([1, 2])
    getMemoryGameGameCards.mockReturnValue(4)
    getMemoryGameIsGameDemonstration.mockReturnValue(false)
  })

  afterEach(() => {
    jest.clearAllMocks()
    cleanup()
  })

  it('Renders without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(component, div)
    ReactDOM.unmountComponentAtNode(div)
  })

  it('Snapshot matchs', () => {
    const { container } = render(component)
    expect(container).toMatchSnapshot()
  })
})
