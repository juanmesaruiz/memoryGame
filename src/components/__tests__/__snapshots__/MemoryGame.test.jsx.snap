// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemoryGame test suite Snapshot matchs 1`] = `
<div>
  <div
    class="memoryGame"
  >
    <section
      class="memoryGame-cards"
    >
      <code
        name="MemoryGameCard"
        props="[object Object]"
      />
      <code
        name="MemoryGameCard"
        props="[object Object]"
      />
      <code
        name="MemoryGameCard"
        props="[object Object]"
      />
      <code
        name="MemoryGameCard"
        props="[object Object]"
      />
    </section>
    <aside
      class="memoryGame-action"
    >
      <button
        class="memoryGame-action-startButton"
      >
        Start game
      </button>
      <div>
        <label>
          Number of cards:
        </label>
        <select
          aria-label="memoryGame-action-select"
          class="memoryGame-action-select"
        >
          <option>
            2
          </option>
          <option>
            3
          </option>
          <option>
            4
          </option>
          <option>
            5
          </option>
          <option>
            6
          </option>
          <option>
            7
          </option>
          <option>
            8
          </option>
        </select>
      </div>
      <div>
        <button
          class="memoryGame-action-goldButton"
        >
          Gold Mode
        </button>
      </div>
      <div
        class="memoryGame-score"
      >
        <p>
          Level: 
          2
        </p>
        <p>
          Best score: 
          <b />
        </p>
      </div>
    </aside>
  </div>
</div>
`;
